### ISSUE TITLE
[HOLD for payment 2023-02-15] [$250] Text is displayed in Times New roman when the site renders first

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12267]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open a new incognito window
2. Open the developer console, network tab
3. Throttle your network connection to Slow 3G
4. Go to NewDot web (production or staging)

## Expected Result:
When the site first renders, you should see the brand fonts displayed in the app.

## Actual Result:
When the site first renders, text is displayed in Times New Roman instead of our brand font.

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.2.21-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
**Expensify/Expensify Issue URL:** 
**Issue reported by:** @roryabraham 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1666946606212909

https://user-images.githubusercontent.com/43996225/198729516-2170d8fb-ae88-4947-a034-ba518cff9f0d.mp4

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/43996225/198729528-b0569a59-5723-4dfe-b309-3c4762e2611c.png">

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by isabelastisser** [comment]
I will review this by EOD.

**Comment by isabelastisser** [comment]
I didn't find time to work on this yesterday, will take care of this by EOD.

**Comment by isabelastisser** [comment]
Hey @kavimuru, I'm unable to reproduce this.

https://user-images.githubusercontent.com/31258516/199372297-ddbc05d0-c36d-4ad6-bfed-dbe47fb71661.mp4

**Comment by roryabraham** [comment]
I'm very consistently able to reproduce this. @isabelastisser you may need to check off `Disable cache` in your network console.

**Comment by roryabraham** [comment]
@puneetlath Going to snag this from you because I'm curious to learn more about the solution. I think it will probably be a simple solution, building off of the technique explained in [this article](https://web.dev/codelab-preload-web-fonts/):

```html
<link rel="preload" href="/assets/Pacifico-Bold.woff2" as="font" type="font/woff2">
``` 

And maybe we need to do that for all our fonts, but I'd like to see that generalized / DRY if possible

**Comment by tienifr** [comment]
# Proposal

## Problem

- This issue occur because of [FOIT](https://fonts.google.com/knowledge/glossary/foit) 
- To fix it, as @roryabraham's suggestion, we'll set preload for font.css

## Solution

In https://github.com/Expensify/App/blob/72fa14d878281b3db1a1fa28330b9fba835a7436/web/index.html#L15 we'll change `rel="stylesheet"` to `rel="preload"`.

```diff
- <link rel="stylesheet" type="text/css" href="/css/fonts.css" />
+ <link rel="preload"  type="text/css" href="/css/fonts.css" />
```

https://user-images.githubusercontent.com/113963320/199467009-51141ab1-e8ab-4db7-a6c7-ebcf2bf976ae.mp4

**Comment by isabelastisser** [comment]
Hey @sobitneupane , can you please review the proposal above? Thanks!

**Comment by sobitneupane** [comment]
@tienifr Is it necessary to preload all the fonts? Performance-wise, It is better to preload absolutely required fonts only.

**Comment by alexandermendes** [comment]
**Proposal**

If performance is a concern, an alternative way of dealing with this issue is to choose web safe fallback fonts that closely mimic the final custom font. It’s the disconnect between the current default and the custom font that’s so jarring, particularly as the fallback is a serif font and the custom font is sans-serif.

So currently the two fonts side by side look like this (taking the landing page example):

<img width="312" alt="image" src="https://user-images.githubusercontent.com/5636273/199801918-87c59ff6-fdf5-47b0-b6cc-0679e531853b.png">

<img width="323" alt="image" src="https://user-images.githubusercontent.com/5636273/199801825-d538a11b-2191-47fd-b667-592960fbbc72.png">

By using a web safe `font-family` as the fallback, of say `'Helvetica Neue', Helvetica, arial, sans-serif`, along with some tweaks to line height and font size and so on we could instead make the fallback look like:

<img width="308" alt="image" src="https://user-images.githubusercontent.com/5636273/199802093-80162d67-df0f-4d81-a37e-fb5dfc9503bc.png">

As we can see the difference between those two, while still present, is far less noticeable. It's a pretty common pattern:

- https://css-tricks.com/how-to-load-fonts-in-a-way-that-fights-fout-and-makes-lighthouse-happy/
- https://www.smashingmagazine.com/2021/05/reduce-font-loading-impact-css-descriptors/#making-your-fallback-font-more-closely-match-your-font

There would be some design choices to make for exactly how to achieve this with your `@font-face` setup, which may not be totally straightforward, but I thought I would throw this one out there to see what you think before digging in for a more refined proposal!

**Comment by alexandermendes** [comment]
As a follow-up, there are probably two overall ways to achieve the above. 

The simple way, that would still be an improvement, would be to not worry about tweaking line heights and font sizes and so on and simply declare fallbacks for each of the custom fonts in the [`_fonts.scss`](https://github.com/Expensify/App/blob/main/docs/_sass/_fonts.scss) file, for example:

```diff
@font-face {
+ font-family: GTAmericaExp-Regular, 'Helvetica Neue', Helvetica, arial, sans-serif;
- font-family: GTAmericaExp-Regular;
  font-weight: 700;
  font-style: normal;
  src: url("https://www.expensify.com/font/GT-America-Standard-Bold.eot") format("embedded-opentype"), url("https://www.expensify.com/font/GT-America-Standard-Bold.woff") format("woff"), url("https://www.expensify.com/font/GT-America-Standard-Bold.woff2") format("woff2");
}
```

That would achieve something like this on initial load (with a slow connection):

https://user-images.githubusercontent.com/5636273/199807009-a7022538-d591-4c98-bb8c-614f28a217c0.mov

Which is definitely an improvement over what we get currently:

https://user-images.githubusercontent.com/5636273/199807157-40ecb943-b921-4d02-a626-0e80975983f8.mov

The more accurate but also more complicated technique (where we also adjust line heights and font sizes and so on) would probably involve some use of the [`FontFace.load()`](https://developer.mozilla.org/en-US/docs/Web/API/FontFace/load) and swapping out the fallbacks with the custom fonts in JS.

**Comment by tienifr** [comment]
> @tienifr Is it necessary to preload all the fonts? Performance-wise, It is better to preload absolutely required fonts only.

 @sobitneupane I think FOUC can occur every page not only on login page, so we need to preload all the fonts to prevent this issue.
For performance, all the fonts are loaded only once and will be cached next time.

**Comment by sobitneupane** [comment]
@alexandermendes The use of fallback fonts as shown in your [video](https://user-images.githubusercontent.com/5636273/199807009-a7022538-d591-4c98-bb8c-614f28a217c0.mov) has shown a good improvement though change is noticeable. I believe use of fallback font will be system-dependent. So, different systems(Mac, Ubuntu,etc.) may have different appearance.

**Comment by sobitneupane** [comment]
> I think FOUC can occur every page not only on login page, so we need to preload all the fonts to prevent this issue.

@tienifr I agree with you as we allow users to open any page directly using deep links.

DO you find any way to make it generalized / DRY as suggested by @roryabraham ?

> And maybe we need to do that for all our fonts, but I'd like to see that generalized / DRY if possible

**Comment by alexandermendes** [comment]
> I believe use of fallback font will be system-dependent. So, different systems(Mac, Ubuntu,etc.) may have different appearance.

Partly true. Although this is why I was referring to webs-safe fonts - i.e. fonts that will be present across a large range of devices (see https://developer.mozilla.org/en-US/docs/Learn/CSS/Styling_text/Fundamentals#web_safe_fonts).

In my example, Arial and sans-serif will be present on virtually every system. Helvetica and Helvetica Neue in that example will be present on 100% of Macs. So with that combo we're talking about maybe 0.1% of devices won't be covered. Not sure how reliable [this source](https://www.cssfontstack.com/) is, say, but what I do know is that we have sites with hundreds of millions of monthly users that use this approach, it works! 

It really depends what is more important to you, do you want to delay the rendering of the page a little but ensure that when it is loaded it uses the correct custom font? Or do you want to let the page load as quickly as possible and accept that there could be some minor change noticeable still between the fallback and the custom font?

**Comment by sobitneupane** [comment]
Waiting @tienifr Response 

> > I think FOUC can occur every page not only on login page, so we need to preload all the fonts to prevent this issue.
> 
> @tienifr I agree with you as we allow users to open any page directly using deep links.
> 
> DO you find any way to make it generalized / DRY as suggested by @roryabraham ?
> 
> > And maybe we need to do that for all our fonts, but I'd like to see that generalized / DRY if possible

**Comment by tienifr** [comment]
I have other solution for you @sobitneupane . We'll preload our font with javascript.
We'll add `script` tag in `header`
``` diff
+ <script>
+           const fonts = [
+               'https://www.expensify.com/font/GT-America-Standard-Light.woff',
+               'https://www.expensify.com/font/GT-America-Standard-Thin.woff',
+               'https://www.expensify.com/font/GT-America-Standard-Regular.woff',
+               'https://www.expensify.com/font/GT-America-Standard-Medium.woff',
+               'https://www.expensify.com/font/GT-America-Standard-Bold.woff',
+               'https://www.expensify.com/font/GT-America-Exp-Mono-Regular.woff',
+               'https://www.expensify.com/font/GT-America-Exp-Mono-Bold.woff',
+           ];
+            fonts.forEach((font,index)=>{
+                const loadFont = new FontFace(index, `url(${font})`);
+                loadFont.load();
+            })
+ </script>
```

**Comment by sobitneupane** [comment]
@tienifr's [proposal1](https://github.com/Expensify/App/issues/12267#issuecomment-1300005810) and [proposal2](https://github.com/Expensify/App/issues/12267#issuecomment-1306621239) both look good to me. We can go with any one of the two.

🎀👀🎀 C+ reviewed

cc: @roryabraham

**Comment by isabelastisser** [comment]
Not overdue, waiting for @roryabraham to review the proposal.

**Comment by isabelastisser** [comment]
bump

**Comment by puneetlath** [comment]
@roryabraham To help us clear out the large backlog of /App bugs, we're putting the spotlight every bug in the repo already than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:  
 
- Decide whether any proposals currently meet our guidelines and can be approved as-is 
- For any that can't, please take this issue internal and treat it as one of your highest priorities 
- If you have any questions, don't hesitate to start a discussion in #bug-zero

**Comment by isabelastisser** [comment]
Bump @roryabraham

**Comment by marcaaron** [comment]
Not @roryabraham  - but great conversation here. I think it would be good to loop in @shawnborton on this one since it touches on the design.

As I understand it, the two current proposals are:

## Option 1:

**Use a sans serif system fallback font to minimize the FOUC impact**

**Pro:** Faster site load.
**Con:** Slightly noticeable compared to loading the actual required fonts, but not as jarring as Times New Roman.

## Option 2:

**Preload the fonts to eliminate FOUC entirely**

**Pro:** No trace of flashing text at all
**Con:** May slow down the page load time a bit

Honestly, I am not sure if we should be too concerned about Option 2 here. The fonts are 50kb woff files and will load pretty fast. I think we can just follow the basic advice [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload) and fix this easily.

**Comment by shawnborton** [comment]
Regardless of what we do, I agree we should use sans-serif system fonts as fall back to minimize FOUC. That being said, Option 2 would be great.

**Comment by tienifr** [comment]
Hi @marcaaron @shawnborton if option 2 is great, we can implement [my above solution](https://github.com/Expensify/App/issues/12267#issuecomment-1306621239) and verify which fonts we need to preload.
Please tell me if we need to implement a PR

**Comment by shawnborton** [comment]
I'll let @marcaaron make the call on that.

**Comment by marcaaron** [comment]
I'm not too clear on the proposal @roryabraham. Are changes required? If so, what are they and can we have @tienifr implement them?

**Comment by isabelastisser** [comment]
bump @roryabraham

**Comment by roryabraham** [comment]
I put up a [draft PR](https://github.com/Expensify/App/pull/12983) with my solution to this issue. It just needs to be fully tested w/ screenshots and then it will be ready for review.

**Comment by roryabraham** [comment]
Putting this on HOLD for https://github.com/Expensify/App/pull/12656

**Comment by isabelastisser** [comment]
not overdue.

**Comment by roryabraham** [comment]
Taking this off hold

**Comment by roryabraham** [comment]
Actually, it's probably a good idea to HOLD on https://github.com/Expensify/App/pull/13714/files#diff-c1f274f56cc349e527b06171a0c50e1393049c0f80ac8ab7df644794d36a2e80

**Comment by roryabraham** [comment]
PR is off HOLD and ready for review.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.67-7 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/12983


If no regressions arise, payment will be issued on 2023-02-15. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@sobitneupane / @roryabraham] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@sobitneupane / @roryabraham] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@sobitneupane / @roryabraham] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@sobitneupane] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@isabelastisser] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by isabelastisser** [comment]
hi @sobitneupane , I invited you to the job in Upwork. Please apply and I will hire you and process the payment ASAP!
Thanks!

**Comment by sobitneupane** [comment]
@isabelastisser As it was an internal PR review, payment amount should be $1000.

**Comment by roryabraham** [comment]
@sobitneupane I think since this PR predates the base price change, $250 is correct.

**Comment by isabelastisser** [comment]
Payment made in Upwork, we are all set! Can I close this?
